service/header/exchange.go

```diff
@@ -0,0 +1,194 @@
+package header
+
+import (
+	"context"
+	"fmt"
+
+	pb "github.com/celestiaorg/celestia-node/service/header/pb"
+	"github.com/celestiaorg/go-libp2p-messenger/serde"
+
+	"github.com/libp2p/go-libp2p-core/host"
+	"github.com/libp2p/go-libp2p-core/network"
+	"github.com/libp2p/go-libp2p-core/peer"
+	"github.com/libp2p/go-libp2p-core/protocol"
+)
+
+var headerExchangeProtocolID = protocol.ID("header-exchange")
+
+// exchange enables sending outbound ExtendedHeaderRequests as well as
+// handling inbound ExtendedHeaderRequests.
+type exchange struct {
+	host host.Host
+	// TODO @renaynay: post-Devnet, we need to remove reliance of Exchange on one bootstrap peer
+	// Ref https://github.com/celestiaorg/celestia-node/issues/172#issuecomment-964306823.
+	peer  *peer.AddrInfo
+	store Store
+}
+
+func newExchange(host host.Host, peer *peer.AddrInfo, store Store) *exchange {
+	ex := &exchange{
+		host:  host,
+		peer:  peer,
+		store: store,
+	}
+	host.SetStreamHandler(headerExchangeProtocolID, ex.requestHandler)
+	return ex
+}
+
+// requestHandler handles inbound ExtendedHeaderRequests.
+func (e *exchange) requestHandler(stream network.Stream) {
+	// unmarshal request
+	pbreq := new(pb.ExtendedHeaderRequest)
+	_, err := serde.Read(stream, pbreq)
+	if err != nil {
+		log.Errorw("reading header request from stream", "err", err.Error())
+		//nolint:errcheck
+		stream.Reset()
+		return
+	}
+	request := &ExtendedHeaderRequest{
+		Origin: pbreq.Origin,
+		Amount: pbreq.Amount,
+	}
+	// route depending on amount of headers requested
+	if request.Amount > 1 {
+		e.handleMultipleHeaderRequest(request, stream)
+	} else {
+		e.handleSingleHeaderRequest(request.Origin, stream) // TODO @renaynay: should we parallelise this?
+	}
```

Comment:
You don't need this if tbh. You can loop around ‘request.Amount’ the same as you do handleMultipleHeaderRequest, so it there would be generalized logic for single and multiple cases. Really, they are identical.